File: backend/app/routers/webhooks.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from datetime import datetime
from ..database import get_db
from ..models import Order, OrderItem
from ..schemas import WebhookRequest, WebhookResponse
from ..ai import classify_email, extract_order_data
import uuid

router = APIRouter(prefix="/api/webhooks", tags=["webhooks"])


def order_to_response(order):
    return {
        "id": order.id,
        "order_number": order.order_number,
        "vendor": order.vendor,
        "customer_name": order.customer_name,
        "status": order.status,
        "location": order.location,
        "expected_date": order.expected_date,
        "notes": order.notes,
        "created_at": order.created_at.isoformat() if order.created_at else "",
        "updated_at": order.updated_at.isoformat() if order.updated_at else "",
        "items": []
    }


@router.post("/order", response_model=WebhookResponse)
async def handle_webhook(body: WebhookRequest, db: AsyncSession = Depends(get_db)):
    if not body.body and not body.snippet:
        raise HTTPException(status_code=400, detail="Missing email content (body or snippet)")
    
    email_content = body.body or body.snippet
    
    classification = await classify_email(body.subject or "", email_content)
    
    if not classification.get("isOrderEmail", False):
        return {
            "message": "Email is not order-related",
            "action": "skipped"
        }
    
    extraction = await extract_order_data(body.subject or "", email_content)
    
    if not extraction.get("extraction_success", False):
        return {
            "message": "Failed to extract order data",
            "action": "failed",
            "classification": classification,
            "extraction": extraction
        }
    
    order_number = extraction.get("order_number")
    if not order_number:
        raise HTTPException(status_code=400, detail="Could not extract order number")
    
    vendor = extraction.get("vendor")
    customer_name = extraction.get("customer_name")
    order_status = extraction.get("order_status", "Ordered")
    delivery_info = extraction.get("delivery_info", {})
    items = extraction.get("items", [])
    
    result = await db.execute(select(Order).where(Order.order_number == order_number))
    existing_order = result.scalar_one_or_none()
    
    if existing_order:
        existing_order.vendor = vendor or existing_order.vendor
        existing_order.customer_name = customer_name or existing_order.customer_name
        existing_order.status = order_status or existing_order.status
        existing_order.location = delivery_info.get("location") or existing_order.location
        existing_order.expected_date = delivery_info.get("expected_date") or existing_order.expected_date
        existing_order.updated_at = datetime.utcnow()
        
        if items:
            await db.execute(OrderItem.__table__.delete().where(OrderItem.order_id == existing_order.id))
            for item in items:
                price = item.get("price")
                if isinstance(price, str):
                    price = float(price.replace("AED", "").replace("AED", "").replace("$", "").strip()) if price else None
                
                order_item = OrderItem(
                    id=str(uuid.uuid4()),
                    order_id=existing_order.id,
                    item_name=item.get("item_name"),
                    quantity=item.get("quantity", 1),
                    price=price,
                    currency=item.get("currency", "AED")
                )
                db.add(order_item)
        
        await db.commit()
        await db.refresh(existing_order)
        
        return {
            "message": "Order updated successfully",
            "action": "updated",
            "order": order_to_response(existing_order),
            "classification": classification,
            "extraction": extraction
        }
    else:
        new_order = Order(
            id=str(uuid.uuid4()),
            order_number=order_number,
            vendor=vendor or "Unknown",
            customer_name=customer_name or "Unknown",
            status=order_status or "Ordered",
            location=delivery_info.get("location", ""),
            expected_date=delivery_info.get("expected_date", "")
        )
        db.add(new_order)
        
        if items:
            for item in items:
                price = item.get("price")
                if isinstance(price, str):
                    price = float(price.replace("AED", "").replace("$", "").strip()) if price else None
                
                order_item = OrderItem(
                    id=str(uuid.uuid4()),
                    order_id=new_order.id,
                    item_name=item.get("item_name"),
                    quantity=item.get("quantity", 1),
                    price=price,
                    currency=item.get("currency", "AED")
                )
                db.add(order_item)
        
        await db.commit()
        await db.refresh(new_order)
        
        return {
            "message": "Order created successfully",
            "action": "created",
            "order": order_to_response(new_order),
            "classification": classification,
            "extraction": extraction
        }

File: backend/src/routes/webhooks.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { getDb } from '../models/database.js';
import { classifyEmail, extractOrderData } from '../services/ai.js';

const router = express.Router();

router.post('/order', async (req, res) => {
  const db = getDb();
  const { subject, body, from, snippet } = req.body;

  if (!body && !snippet) {
    return res.status(400).json({ error: 'Missing email content (body or snippet)' });
  }

  const emailContent = body || snippet;

  try {
    const classification = await classifyEmail(subject || '', emailContent);

    if (!classification.isOrderEmail) {
      return res.json({
        message: 'Email is not order-related',
        classification,
        action: 'skipped'
      });
    }

    const extraction = await extractOrderData(subject || '', emailContent);

    if (!extraction.extraction_success) {
      return res.json({
        message: 'Failed to extract order data',
        classification,
        extraction,
        action: 'failed'
      });
    }

    const { 
      order_number, 
      vendor, 
      customer_name, 
      order_status, 
      delivery_info, 
      items, 
      order_total 
    } = extraction;

    if (!order_number) {
      return res.status(400).json({ error: 'Could not extract order number' });
    }

    const existingOrder = db.prepare('SELECT * FROM orders WHERE order_number = ?').get(order_number);

    if (existingOrder) {
      const updateOrder = db.prepare(`
        UPDATE orders 
        SET vendor = ?, customer_name = ?, status = ?, 
            location = ?, expected_date = ?, updated_at = datetime('now')
        WHERE id = ?
      `);
      
      updateOrder.run(
        vendor ?? existingOrder.vendor,
        customer_name ?? existingOrder.customer_name,
        order_status ?? existingOrder.status,
        delivery_info?.location ?? existingOrder.location,
        delivery_info?.expected_date ?? existingOrder.expected_date,
        existingOrder.id
      );

      if (items && items.length > 0) {
        db.prepare('DELETE FROM order_items WHERE order_id = ?').run(existingOrder.id);
        
        const insertItem = db.prepare(`
          INSERT INTO order_items (id, order_id, item_name, quantity, price, currency)
          VALUES (?, ?, ?, ?, ?, ?)
        `);
        
        for (const item of items) {
          const price = item.price ? parseFloat(item.price.replace(/[^0-9.-]+/g, '')) : null;
          insertItem.run(uuidv4(), existingOrder.id, item.item_name, item.quantity || 1, price, item.currency || 'AED');
        }
      }

      const updatedOrder = db.prepare('SELECT * FROM orders WHERE id = ?').get(existingOrder.id);
      const orderItems = db.prepare('SELECT * FROM order_items WHERE order_id = ?').all(existingOrder.id);

      return res.json({ 
        message: 'Order updated successfully', 
        order: { ...updatedOrder, items: orderItems },
        action: 'updated',
        classification,
        extraction
      });
    } else {
      const id = uuidv4();
      
      const insertOrder = db.prepare(`
        INSERT INTO orders (id, order_number, vendor, customer_name, status, location, expected_date)
        VALUES (?, ?, ?, ?, ?, ?, ?)
      `);
      
      insertOrder.run(
        id,
        order_number,
        vendor || 'Unknown',
        customer_name || 'Unknown',
        order_status || 'Ordered',
        delivery_info?.location || '',
        delivery_info?.expected_date || ''
      );

      if (items && items.length > 0) {
        const insertItem = db.prepare(`
          INSERT INTO order_items (id, order_id, item_name, quantity, price, currency)
          VALUES (?, ?, ?, ?, ?, ?)
        `);
        
        for (const item of items) {
          const price = item.price ? parseFloat(item.price.replace(/[^0-9.-]+/g, '')) : null;
          insertItem.run(uuidv4(), id, item.item_name, item.quantity || 1, price, item.currency || 'AED');
        }
      }

      const newOrder = db.prepare('SELECT * FROM orders WHERE id = ?').get(id);
      const orderItems = db.prepare('SELECT * FROM order_items WHERE order_id = ?').all(id);

      return res.status(201).json({ 
        message: 'Order created successfully', 
        order: { ...newOrder, items: orderItems },
        action: 'created',
        classification,
        extraction
      });
    }
  } catch (error) {
    console.error('Webhook error:', error);
    res.status(500).json({ error: error.message });
  }
});

export default router;
